modules/dists/BlockDist.chpl

```diff
@@ -869,24 +925,128 @@ iter BlockImpl.activeTargetLocales(const space : domain = boundingBox) {
   }
 }
 
-proc type Block.createDomain(dom: domain) {
-  return dom dmapped Block(dom);
+// create a domain over an existing Block Distribution
+proc Block.createDomain(dom: domain(?)) {
+  return dom dmapped this;
+}
+
+// create a domain over an existing Block Distribution constructed from a series of ranges
+proc Block.createDomain(rng: range(?)...) {
+  return this.createDomain({(...rng)});
+}
+
+// create a domain over a Block Distribution
+proc type Block.createDomain(dom: domain(?), targetLocales: [] locale = Locales) {
+  return dom dmapped Block(dom, targetLocales);
+}
+
+// create a domain over a Block Distribution constructed from a series of ranges
+proc type Block.createDomain(rng: range(?)..., targetLocales: [] locale = Locales) {
+  return createDomain({(...rng)}, targetLocales);
 }
 
-proc type Block.createDomain(rng: range...) {
+proc type Block.createDomain(rng: range(?)...) {
   return createDomain({(...rng)});
 }
 
-proc type Block.createArray(dom: domain, type eltType) {
-  var D = createDomain(dom);
+// create an array over a Block Distribution, default initialized
+proc type Block.createArray(
+  dom: domain(?),
+  type eltType,
+  targetLocales: [] locale = Locales
+) {
+  var D = createDomain(dom, targetLocales);
   var A: [D] eltType;
   return A;
 }
 
-proc type Block.createArray(rng: range..., type eltType) {
+// create an array over a Block Distribution, initialized with the given value or iterator
+@unstable("'Block.createArray' with an 'initExpr' formal is unstable and may change in a future release")
+proc type Block.createArray(
+  dom: domain(?),
+  type eltType,
+  initExpr: ?t,
+  targetLocales: [] locale = Locales
+) where isSubtype(t, _iteratorRecord) || isCoercible(t, eltType)
+{
+  var D = createDomain(dom, targetLocales);
+  var A: [D] eltType;
+  A = initExpr;
+  return A;
+}
+
+// create an array over a Block Distribution, initialized from the given array
+@unstable("'Block.createArray' with an 'initExpr' formal is unstable and may change in a future release")
+proc type Block.createArray(
+  dom: domain(?),
+  type eltType,
+  initExpr: [?arrayDom] ?arrayEltType,
+  targetLocales: [] locale = Locales
+) where dom.rank == arrayDom.rank && isCoercible(arrayEltType, eltType)
+{
+  for (d, ad, i) in zip(dom.dims(), arrayDom.dims(), 0..) do
+    if d.size != ad.size then halt("Domain size mismatch in 'Block.createArray' dimension " + i:string);
```

Comment:
I know a bit after the fact, but I'm wondering if we should only be doing this check if `boundsChecking`? Also, is there a reason this is a `for` and not a `forall`?

Reply:
Ah, I suppose since it's just over the `dims`, that isn't likely to be expensive, so I could see the `forall` overhead not being worth it, but I do still wonder if we only want to do this if bounds checking is enabled.

Reply:
> I know a bit after the fact, but I'm wondering if we should only be doing this check if boundsChecking?

I did not think about that, but you're probably right. I can make the change in another PR soon.

> Also, is there a reason this is a for and not a forall?

I was thinking that since arrays rarely exceed a handful of dimensions (afaik), it felt too heavy weight to put a `forall` there and potentially spawn new tasks for that check.

Reply:
That makes sense to me, thanks Jeremiah, hope the post-facto review wasn't too annoying haha.

Reply:
No worries, I appreciate the feedback, thanks!